comms/src/message/message_envelope.rs

```diff
@@ -0,0 +1,366 @@
+        }
+
+        let peer_signature = peer_signature(node_identity.secret_key.clone(), &body)?;
+
+        let header = MessageEnvelopeHeader {
+            public_key: node_identity.identity.public_key.clone(),
+            signature: peer_signature,
+            flags,
+        };
+
+        Ok(Self::new(
+            header.to_binary().map_err(MessageError::MessageFormatError)?,
+            body,
+        ))
+    }
+
+    /// Returns the frame that is expected to be version frame
+    pub fn version_frame(&self) -> &Frame {
+        &self.frames[0]
+    }
+
+    /// Returns the frame that is expected to be header frame
+    pub fn header_frame(&self) -> &Frame {
+        &self.frames[1]
+    }
+
+    /// Returns the [MessageEnvelopeHeader] deserialized from the header frame
+    pub fn deserialize_header(&self) -> Result<MessageEnvelopeHeader, MessageError> {
+        MessageEnvelopeHeader::from_binary(self.header_frame()).map_err(Into::into)
+    }
+
+    /// Returns the frame that is expected to be body frame
+    pub fn body_frame(&self) -> &Frame {
+        &self.frames[2]
+    }
+
+    /// Consumes the envelope and returns the body frame
+    pub fn into_body_frame(mut self) -> Frame {
+        self.frames.remove(2)
+    }
+
+    /// Returns the version of this envelope
+    pub fn version(&self) -> u8 {
+        let frame = self.version_frame();
+        // If version not specified, it's assumed to be 0
+        if frame.len() == 0 {
+            return 0;
+        }
+        frame[0]
+    }
+
+    /// Returns a decrypted version of the body frame
+    pub fn decrypted_body_frame(
+        &self,
+        secret_key: &CommsSecretKey,
+        source_public_key: &CommsPublicKey,
+    ) -> Result<Frame, MessageError>
+    {
+        let ecdh_shared_secret = crypt::generate_ecdh_secret(secret_key, source_public_key);
+        let decrypted_frame =
+            crypt::decrypt(&ecdh_shared_secret, self.body_frame()).map_err(MessageError::CipherError)?;
+        Ok(decrypted_frame)
+    }
+
+    /// Returns the decrypted and deserialized Message from the body frame
+    pub fn deserialize_encrypted_body(
+        &self,
+        secret_key: &CommsSecretKey,
+        source_public_key: &CommsPublicKey,
+    ) -> Result<Message, MessageError>
+    {
+        Message::from_binary(&self.decrypted_body_frame(secret_key, source_public_key)?).map_err(Into::into)
+    }
+
+    /// This struct is consumed and the contained FrameSet is returned.
+    pub fn into_frame_set(self) -> FrameSet {
+        self.frames
+    }
+}
+
+impl TryFrom<FrameSet> for MessageEnvelope {
+    type Error = MessageError;
+
+    /// Returns a MessageEnvelope from a FrameSet
+    fn try_from(frames: FrameSet) -> Result<Self, Self::Error> {
+        if frames.len() != FRAMES_PER_MESSAGE {
+            return Err(MessageError::InvalidMultipartMessageLength);
+        }
+
+        Ok(MessageEnvelope { frames })
+    }
+}
+
+//#[cfg(test)]
```

Comment:
Can these tests be added back?

Reply:
I've added tests for the impls on the new message types. Some of these tests won't apply, specifically MessageEnvelope won't be needed.

Reply:
I'll add back the applicable tests in a subsequent PR